source/common/network/lc_trie.h

```diff
@@ -0,0 +1,487 @@
+                               extractBits<IpType, address_size>(0, length_, ip_) ==
+                                   extractBits<IpType, address_size>(0, length_, other.ip_)));
+    }
+
+    // The address represented either in uint32_t or uint128.
+    IpType ip_;
+    // Length of the cidr range.
+    int length_;
+    // TODO(ccaraman): Support more than one tag per entry.
+    // Tag for this entry.
+    std::string tag_;
+  };
+
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)> class LcTrieInternal {
+  public:
+    /**
+     * Level Compressed Trie (LC-Trie) implementation for IP(IPv4 or IPv6) prefixes.
+     *
+     * The C++ implementation of the LcTrie found at
+     * https://github.com/beevek/libkrb/blob/master/krb/lc_trie.hpp was used as a reference.
+     * The implemenation above is based off of S. Nilson's code at
+     * http://www.csc.kth.se/~snilsson/software/router/C/
+     *
+     * Note: This trie handles Ipv4 and Ipv6 separately.
+     * Note: The trie can only support up 524288(2^19) nodes. Please see LcNode for further
+     * explanation.
+     *
+     * @param fill_factor supplies the fraction of completeness to use when considering
+     *        if a subtrie should be compressed.
+     * @param root_branching_factor supplies the branching factor at the root. This is
+     *        an optimization as suggested in the paper.
+     * @param tag_data supplies the ip tag and the prefixes for associated with that tag.
+     */
+    LcTrieInternal(std::vector<IpPrefixes<IpType>>& tag_data, double fill_factor = 0.5,
+                   uint32_t root_branching_factor = 0);
+
+    /**
+     * Search the LC Trie for a prefix that has the input, ip, in range.
+     * @param  ip supplies the IP address searching for a prefix in string format.
+     * @return tag from the prefix that encompasses the input. An empty string is returned
+     *         if no prefix in the LC Trie exists.
+     */
+    std::string search(const IpType& ip_address) const;
+
+  private:
+    /**
+     * Builds the Level Compresesed Trie, by first sorting the base vector, removing nested prefixes
+     * and duplicated and then recursively builds the trie by invoking buildRecursively().
+     */
+    void buildLcTrie(std::vector<IpPrefixes<IpType>>& tag_data) {
+      if (tag_data.empty()) {
+        return;
+      }
+
+      // The node used for the trie uses the last 20 bits of uint32_t. The trie cannot support
+      // an input greater than 524288(2^19) nodes.
+      // TODO(ccaraman): Add a test case for this.
+      if (tag_data.size() > (1 << 19)) {
+        throw EnvoyException(fmt::format(
+            "Size of the input vector({0}) for LC-Trie is larger than the max size supported(2^19)",
+            tag_data.size()));
+      }
+
+      std::sort(tag_data.begin(), tag_data.end());
+      ip_prefixes_.push_back(tag_data[0]);
+
+      // Remove duplicate entries and nested prefixes.
```

Comment:
Does the original LC trie code work with nested prefixes? I know the corresponding paper didn't use any nested prefixes in its examples, but it isn't clear to me if the design can handle nested prefixes without modifications.

Reply:
I'll update the comment here and for the class and method definition. Currently, the code will throw an exception if there are nested prefixes. I have a todo to make the exception show which prefixes are nested. 

The future implementation for nested prefixes won't modify the algorithm or the underlying LcNode. There are two options to handle it: 
1. Have a sorted list of the nested prefixes 
2. Use another lc-trie for the nested prefixes
The most general prefix will have a way of accessing one of those options once it is implemented.

Reply:
FWIW, I would do the implementation that was done in the original C implementation. IMO it's very easy to understand and will perform very well (linked list of prefixes, accessed from trie node). It will also perform well for our case where 0 prefixes will be by far the most common case and 1 prefix will be the next most common case by a huge margin.